packages/hub/src/lib/parse-safetensors.ts

```diff
@@ -0,0 +1,142 @@
+import { HUB_URL } from "../consts";
+import { Counter } from "../utils/Counter";
+
+const SINGLE_FILE = "model.safetensors";
+const INDEX_FILE = "model.safetensors.index.json";
+
+type FileName = string;
+
+type TensorName = string;
+type Dtype = "F64" | "F32" | "F16" | "BF16" | "I64" | "I32" | "I16" | "I8" | "U8" | "BOOL";
+
+interface TensorInfo {
+	dtype: Dtype;
+	shape: number[];
+	data_offsets: [number, number];
+}
+
+type FileHeader = Record<TensorName, TensorInfo> & {
+	__metadata__: Record<string, string>;
+};
+
+interface IndexJson {
+	dtype?: string;
+	/// ^there's sometimes a dtype but it looks inconsistent.
+	metadata?: Record<string, string>;
+	/// ^ why the naming inconsistency?
+	weight_map: Record<TensorName, FileName>;
+}
+
+type ShardedHeaders = Record<FileName, FileHeader>;
+
+type ParseFromRepo =
+	| {
+			sharded: false;
+			header: FileHeader;
+	  }
+	| {
+			sharded: true;
+			index: IndexJson;
+			headers: ShardedHeaders;
+	  };
+
+async function parseSingleFile(url: URL): Promise<FileHeader> {
+	const bufLengthOfHeaderLE = await (
+		await fetch(url, {
+			headers: {
+				Range: "bytes=0-7",
+			},
+		})
+	).arrayBuffer();
+	const lengthOfHeader = new DataView(bufLengthOfHeaderLE).getBigUint64(0, true);
+	/// ^little-endian
+	const header: FileHeader = await (
+		await fetch(url, {
+			headers: {
+				Range: `bytes=8-${7 + Number(lengthOfHeader)}`,
+			},
+		})
+	).json();
+	/// no validation for now, we assume it's a valid FileHeader.
+	return header;
+}
+
+async function parseShardedIndex(url: URL): Promise<{ index: IndexJson; headers: ShardedHeaders }> {
+	const index: IndexJson = await (await fetch(url)).json();
+	/// no validation for now, we assume it's a valid IndexJson.
+
+	const shardedMap: ShardedHeaders = {};
+	const filenames = [...new Set(Object.values(index.weight_map))];
+	await Promise.all(
+		filenames.map(async (filename) => {
+			const singleUrl = new URL(url.toString().replace(INDEX_FILE, filename));
+			shardedMap[filename] = await parseSingleFile(singleUrl);
+		})
+	);
+	return { index, headers: shardedMap };
+}
+
+async function doesFileExistOnHub(url: URL): Promise<boolean> {
+	const res = await fetch(url, {
+		method: "HEAD",
+		redirect: "manual",
+		/// ^do not follow redirects to save some time
+	});
+	return res.status >= 200 && res.status < 400;
+}
+
+export async function parseSafetensorsFromModelRepo(params: {
+	modelId: string;
+	hubUrl?: string;
+}): Promise<ParseFromRepo> {
```

Comment:
```suggestion
export async function parseSafetensorsMetadata(params: {
  repo: RepoDesignation; // can just be a model id
	hubUrl?: string;
	credentials?: Credentials;
	revision?: string;
}): Promise<ParseFromRepo> {
```

For consistency with the rest of the exports. We can add checks to make sure that `repo` is a model:

```ts
const repoId = toRepoId(params.repo);

if (repoId.type !== "model) {
  throw new TypeError("...");
}
```

Reply:
0ad1d27209d4f02dfc4d83d52e64a7765ca42b1e